File: 2023/Go/day1/day1.go
// Go 1.21.6

package main

import (
	"bufio"
	"errors"
	"fmt"
	"log"
	"os"
	"strconv"
	"unicode"
)

func main() {
	log.SetPrefix("advent of code 2023 day1:")
	log.SetFlags(0)

	var filepath string = "../../inputs/day1.txt"
	file, err := os.Open(filepath)

	if err != nil {
		error_message := fmt.Sprintf("Cannot open file at path \"%v\"", filepath)
		log.Fatal(errors.New(error_message))
	}

	var lines []string
	var total int32 = 0
	var scanner *bufio.Scanner = bufio.NewScanner(file)

	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}

	file.Close()

	for _, line := range lines {
		var numbers []byte

		for i := 0; i < len(line); i++ {
			var character byte = line[i]

			if unicode.IsDigit(rune(character)) {
				numbers = append(numbers, character)
			}
		}

		var number int32 = get_number(numbers)
		total += number
	}

	fmt.Printf("total = %d\n", total)
}

func get_number(numbers []byte) int32 {
	if len(numbers) >= 2 {
		var numeric string = fmt.Sprintf("%c%c", numbers[0], numbers[len(numbers)-1])

		if number, err := strconv.ParseInt(numeric, 10, 32); err == nil {
			return int32(number)
		}
	} else if len(numbers) == 1 {
		var numeric string = fmt.Sprintf("%c%c", numbers[0], numbers[0])

		if number, err := strconv.ParseInt(numeric, 10, 32); err == nil {
			return int32(number)
		}
	}

	return 0
}

File: 2023/PHP/day1.php
<?php
// PHP 8.3.1

$filepath = "../inputs/day1.txt";
$file = fopen($filepath, "r");
$total = 0;

while (($line = fgets($file)) !== false) {
    $numbers = [];

    for ($i = 0; $i < strlen($line); $i++) {
        $char = $line[$i];

        if (is_numeric($char)) {
            array_push($numbers, $char);
        }
    }

    var_dump($numbers);

    if (count($numbers) >= 2) {
        $number = $numbers[0] . $numbers[count($numbers) - 1];
        $total += (int)$number;
    } else if (count($numbers) == 1) {
        $number = $numbers[0] . $numbers[0];
        $total += (int)$number;
    }
}

echo "$total\n";

fclose($file);